pyquil/api/_config.py

```diff
@@ -93,65 +135,164 @@ class PyquilConfig(object):
+    def engage_cmd(self):
+        return self._env_or_config_or_default(**self.ENGAGE_CMD)
+
+    @property
+    def engagement(self):
+        return self.get_engagement()
 
     @property
     def forest_url(self):
         return self._env_or_config_or_default(**self.FOREST_URL)
 
     @property
-    def engage_cmd(self):
-        return self._env_or_config_or_default(**self.ENGAGE_CMD)
+    def qmi_auth_token_path(self):
+        return path.expanduser(self._env_or_config_or_default(**self.QMI_AUTH_TOKEN_PATH))
+
+    @property
+    def qcs_auth_headers(self):
+        if self.user_auth_token is not None:
+            return {'Authorization': 'Bearer %s' % self.user_auth_token['access_token']}
+        if self.qmi_auth_token is not None:
+            return {'X-QMI-AUTH-TOKEN': self.qmi_auth_token['access_token']}
+        return {}
+
+    @property
+    def qcs_url(self):
+        return self._env_or_config_or_default(**self.QCS_URL)
+
+    @property
+    def qpu_compiler_url(self):
+        return self._env_or_config_or_default(**self.QPU_COMPILER_URL)
 
     @property
     def qpu_url(self):
         return self._env_or_config_or_default(**self.QPU_URL)
 
+    @property
+    def quilc_url(self):
+        return self._env_or_config_or_default(**self.QUILC_URL)
+
     @property
     def qvm_url(self):
         return self._env_or_config_or_default(**self.QVM_URL)
 
+    def update_user_auth_token(self, user_auth_token):
+        self.user_auth_token = user_auth_token
+        with open(self.user_auth_token_path, 'w') as f:
+            json.dump(user_auth_token, f)
+
+    def update_qmi_auth_token(self, qmi_auth_token):
+        self.qmi_auth_token = qmi_auth_token
+        with open(self.qmi_auth_token_path, 'w') as f:
+            json.dump(qmi_auth_token, f)
+
     @property
-    def quilc_url(self):
-        return self._env_or_config_or_default(**self.QUILC_URL)
+    def user_auth_token_path(self):
+        return path.expanduser(self._env_or_config_or_default(**self.USER_AUTH_TOKEN_PATH))
 
     @property
-    def qpu_compiler_url(self):
-        return self._env_or_config_or_default(**self.QPU_COMPILER_URL)
+    def user_id(self):
+        return self._env_or_config_or_default(**self.USER_ID)
+
+    def assert_valid_auth_credential(self) -> None:
+        """
+        assert_valid_auth_credential will check to make sure the user has a valid
+        auth credential configured. This assertion is made lazily - it is called
+        only after the user has received a 401 or 403 from forest server. See
+        _base_connection.py::ForestSession#_refresh_auth_token.
+        """
+        if self.user_auth_token is None and self.qmi_auth_token is None:
+            raise UserMessageError(
+                f'Your configuration does not have valid authentication credentials. \
+Please visit {self.qcs_url}/auth/token to download credentials \
+and save to {self.user_auth_token_path}.')
```

Comment:
you can do this without `\` by just hitting enter to break the f-strings up, yeah?